app/src/main/java/io/appium/settings/Settings.java

```diff
@@ -70,12 +79,104 @@ public void onCreate(Bundle savedInstanceState) {
+                            RecorderService.ACTION_RECORDING_FILENAME);
+
+                    if (recordingFilename == null || recordingFilename.isEmpty()
+                            || !recordingFilename.endsWith(".mp4")) {
+                        recordingFilename = "AppiumScreenRecord.mp4";
+                    }
+
+                    recordingOutputPath = getExternalFilesDir(null).getAbsolutePath()
+                            + File.separator + recordingFilename;
+                    Intent returnedIntent = new Intent();
+                    returnedIntent.putExtra("output", recordingOutputPath);
+                    setResult(Activity.RESULT_OK, returnedIntent);
+
+                    // start record
+                    final MediaProjectionManager manager
+                            = (MediaProjectionManager) getSystemService(
+                                    Context.MEDIA_PROJECTION_SERVICE);
+                    final Intent permissionIntent = manager.createScreenCaptureIntent();
+
+                    startActivityForResult(permissionIntent,
+                            RecorderService.REQUEST_CODE_SCREEN_CAPTURE);
+                } else {
+                    finishActivity();
+                }
+            } else if (openingIntent.getAction().equals(RecorderService.ACTION_RECORDING_STOP)
+                    && isHigherThanP()) {
+                // stop record
+                final Intent intent = new Intent(this, RecorderService.class);
+                intent.setAction(RecorderService.ACTION_RECORDING_STOP);
+                startService(intent);
+
+                finishActivity();
+            } else {
+                finishActivity();
+            }
+        }
+    }
+
+    private void finishActivity() {
         Log.d(TAG, "Closing the app");
         Handler handler = new Handler();
         handler.postDelayed(Settings.this::finish, 1000);
     }
 
-    private void registerSettingsReceivers(List<Class<? extends BroadcastReceiver>> receiverClasses) {
+    private boolean checkRecordingPermission() {
+        // Check if we have required permission
+        int permission = ActivityCompat.checkSelfPermission(getApplicationContext(),
+                Manifest.permission.WRITE_EXTERNAL_STORAGE);
+        int permissionAudio = ActivityCompat.checkSelfPermission(getApplicationContext(),
+                Manifest.permission.RECORD_AUDIO);
+
+        return permission == PackageManager.PERMISSION_GRANTED
+                && permissionAudio == PackageManager.PERMISSION_GRANTED;
+    }
+
+    private boolean isHigherThanP() {
+        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q;
+    }
+
+    @Override
+    protected void onActivityResult(final int requestCode, final int resultCode, final Intent data)
+    {
+        super.onActivityResult(requestCode, resultCode, data);
+        if (RecorderService.REQUEST_CODE_SCREEN_CAPTURE == requestCode) {
+            if (resultCode == Activity.RESULT_OK) {
```

Comment:
we could use `&&` instead of nested `if`s

Reply:
done